feat(geocoding): added geocoding documentation (#93)
MapColonies/developer-portal: docs/MapColonies/vector/Services/geocoding/README.md

Reviewer: Please add examples for client common usage. This is important so users can understand and get ideas of how to use the service as an implementation.
This can come in the next version and isn't mandatory for the first version in the portal.

Reviewer: Please reorder examples section as we discussed.

---
id: vector-geocoding-service
slug: info
title: Geocoding 📍
description: Search engine for locations and Control Language
tags:
  - Vector
  - API
  - Geocoding
---

# Geocoding 📍

:::info
Click [here](/docs/MapColonies/vector/Services/geocoding/api) for the OpenAPI
:::

## TL;DR

Search for places, control tiles, routes and items. Ability to convert from WGS 84 Coordinates to Control Tile/ US Army MGRS. 

## Terminology

### What is the Control layer?

:::tip
Check [`Control Search request examples`](#control-examples).
:::

MapColonies has its own Control Reference System. Like the US Army MGRS, we divided our user's area of interest to Tiles. Each tile is 10kmX10km and has Sub-Tiles which are 1kmX1km. Each tile has 100 sub-tiles. <br/>
A tile's name is exactly 3-letters, while a Sub-Tile is a 2-digit number. <br/>

> <details>

Reviewer: Maybe add a short explanation for these responses, or in my opinion just write examples as plain text because we have examples at the end of this document.

            "limit": 5
        },
        "response": {
            "results_count": 1,
            "max_score": 3.588634,
            "match_latency_ms": 5
        }
    },
    "features": [
        {
            "type": "Feature",
            "geometry": {
                "coordinates": [ [ [ 12.439530324602458, 41.93031190061167 ],
                        [12.439646505602468, 41.9393328566117 ],
                        [ 12.429158378602494, 41.939432381611695 ],
                        [ 12.429043189602453, 41.930411390611695 ],
                        [ 12.439530324602458, 41.93031190061167 ] ] ],
                "type": "Polygon"
            },
            "properties": {
                "SUB_TILE_ID": "65",
                "TILE_NAME": "RIT",
                "LAYER_NAME": "CONTROL.SUB_TILES",
                "TYPE": "SUB_TILE",
                "matches": [ { "layer": "CONTROL.SUB_TILES", "source": "control_gil_v5_test", "source_id": [] } ],
                "names": { "default": [ "65" ], "display": "Tile RIT Sub Tile 65" },
                "score": 3.588634
            }
        }
    ]
}
```
</details>

### Understanding the Control language

> Routes are roads and streams. Each Route has Control Points. Control Points are points scattered on the route. They are usefull when end-user reports it's location on the route. Its used to "pin-point" end-user location.<br/>

Reviewer: ```suggestion
Routes are roads and streams. Each Route has Control Points. Control Points are points scattered on the route. They are usefull when an end-user reports his location on the route. It is used to "pin-point" the end-user's location.<br/>
```

> Route name is just a string (in any length). For main roads, the route's name will be unique. For minor roads the name might be reused in other areas of interest. <br/>

Reviewer: ```suggestion
A route's name is a string (in any length). For main roads, the route's name will be unique. For minor roads the name might be reused in other areas of interest. <br/>
```

> Route name is just a string (in any length). For main roads, the route's name will be unique. For minor roads the name might be reused in other areas of interest. <br/>

Reviewer: Maybe add an example as text for the reuse of `minor` road names.

> Route's Control point is a 3-digit number.

Reviewer: ```suggestion
 A route's Control point is a 3-digit number.
```

> <details>

Reviewer: Same here, when giving a response example please elaborate on what the user is about to see and the context.

>     <summary>JSON Schema of Route Response</summary>

Reviewer: Json schema?

```json
{
    "type": "FeatureCollection",
    "geocoding": {
        "version": "0.1.0",
        "query": {
            "command_name": "olimpiade",
            "disable_fuzziness": false,
            "limit": 5
        },
        "response": {
            "results_count": 1,
            "max_score": 3.2834144,
            "match_latency_ms": 16
        }
    },
    "features": [
        {
            "type": "Feature",
            "geometry": {
                "coordinates": [ [ 12.472759211857749, 41.932073169074016 ],
                    [ 12.475880836863752, 41.93239148035457 ] ],
                "type": "LineString"
            },
            "properties": {
                "OBJECT_COMMAND_NAME": "olimpiade",
                "LAYER_NAME": "CONTROL.ROUTES",
                "ENTITY_HEB": "route",
                "TYPE": "ROUTE",
                "matches": [ { "layer": "CONTROL.ROUTES", "source": "control_gil_v5_test", "source_id": [] } ],
                "names": { "default": [ "olimpiade" ], "display": "Route olimpiade" },
                "score": 3.2834144
            }
        }
    ]
}
```
</details>

<details>

>     <summary>JSON Schema of Route's Control Point Response</summary>

Reviewer: Same here.

```json
{
    "type": "FeatureCollection",
    "geocoding": {
        "version": "0.1.0",
        "query": {
            "command_name": "olimpiade",
            "control_point": "111",
            "disable_fuzziness": true,
            "limit": 5
        },
        "response": {
            "results_count": 1,
            "max_score": 1.89712,
            "match_latency_ms": 7
        }
    },
    "features": [
        {
            "type": "Feature",
            "geometry": {
                "coordinates": [ 12.475638293442415, 41.932360642739155 ],
                "type": "Point"
            },
            "properties": {
                "OBJECT_COMMAND_NAME": "111",
                "TIED_TO": "olimpiade",
                "ENTITY_HEB": "control point",
                "LAYER_NAME": "CONTROL_GIL_GDB.CTR_CONTROL_POINT_CROSS_N",
                "TYPE": "ITEM",
                "matches": [ { "layer": "CONTROL_GIL_GDB.CTR_CONTROL_POINT_CROSS_N", "source": "control_gil_v5_test", "source_id": [] } ],
                "names": { "default": [ "111" ], "display": "Route olimpiade Control Point 111" },
                "score": 1.89712
            }
        }
    ]
}
```
</details>

> Items are simply buildings\ items in the field. Item's name consists 4-digit number. It might be reused in different tiles! <br/>

Reviewer: ```suggestion
Items are simply buildings \ items in the field. An item's name is a 4-digit number. Notice that it is not unique and might be reused in different tiles!<br/>
```

> Items are simply buildings\ items in the field. Item's name consists 4-digit number. It might be reused in different tiles! <br/>

Reviewer: Maybe make the notice as a banner?

                        [ 12.432648028923637, 41.93209008126263 ],
                        [ 12.43295235249525, 41.93189198298137 ],
                        [ 12.435105441764364, 41.93235609798671 ],
                        [ 12.435516278586334, 41.93274663116725 ],
                        [ 12.43637599267521, 41.93308056343986 ],
                        [ 12.436026020567567, 41.93386161853829 ],
                        [ 12.432792582620323, 41.9327692706986 ] ] ],
                "type": "Polygon"
            },
            "properties": {
                "OBJECT_COMMAND_NAME": "1234",
                "TILE_NAME": "RIT",
                "SUB_TILE_ID": "37",
                "LAYER_NAME": "CONTROL.ITEMS",
                "ENTITY_HEB": "hotel",
                "TYPE": "ITEM",
                "matches": [ { "layer": "CONTROL.ITEMS", "source": "control_gil_v5_test", "source_id": [] } ],
                "names": { "default": [ "1234" ], "display": "Item 1234" },
                "score": 2.664375
            }
        }
    ]
}
```
</details>

## Location Name Based Search: 
:::tip
Check [`Location Search request examples`](#location-search-examples).
:::

> We have created a location search engine. In the query parameter simple search a place that you want to find. For example, if you search for `White House, Washington DC` you will get matching results for your query. <br/>

Reviewer: ```suggestion
We have created a location search engine. In the query parameter simply search for a place that you want to find. For example, if you search for `White House, Washington DC` you will get matching results for your query. <br/>
```

> You can also search for supported regions and sources so you can filter the results for a specific region/ specific source (it will result results from all regions and sources by default). <br/>

Reviewer: ```suggestion
You can also search for supported regions and sources in order to filter the results for a specific region or source (by default it will return all regions and sources). <br/>
```

                        [ -73.76213624656812, 40.65309424493557 ],
                        [ -73.78181122379466, 40.66270746643491 ],
                        [ -73.79106514121902, 40.66438330508498 ],
                        [ -73.7957754685564, 40.665205399216404 ],
                        [ -73.79856831750864, 40.66283394629252 ],
                        [ -73.80390390953731, 40.66175885985783 ],
                        [ -73.8073637074931, 40.66039916690434 ],
                        [ -73.8109068740743, 40.66074699797244 ],
                        [ -73.81278266814672, 40.66039916690434 ] ] ]
            },
            "properties": {
                "score": 2.2699702,
                "matches": [ { "layer": "osm_airports", "source": "OSM", "source_id": [ "03ed6d97-fc81-4340-b68a-11993554eef1" ] } ],
                "names": { "en": [ "JFK Airport" ], "fr": [ "Aeropuerto JFK" ], "default": [ "JFK" ], "display": "JFK Airport" },
                "placetype": "transportation",
                "sub_placetype": "airport",
                "regions": [ { "region": "USA", "sub_region_names": [ "New York" ] } ]
            }
        }
    ]
}
```
</details>

## Conversions

> :::tip

Reviewer: Maybe move the tips to the end of the section? (For all sections).

Check [`Conversions request examples`](#conversions-1).
:::

> You can convert WGS84 coordinates to 2 grids: Mapcolonies Control Grid/ US Army MGRS. In order to choose your target grid you pass `target_grid` query parameter.<br/>

Reviewer: ```suggestion
You can convert `WGS84` coordinates to two grids, the `Mapcolonies Control Grid` and `US Army MGRS`. In order to choose your target grid you pass the `target_grid` query parameter.<br/>
```

> You can convert WGS84 coordinates to 2 grids: Mapcolonies Control Grid/ US Army MGRS. In order to choose your target grid you pass `target_grid` query parameter.<br/>

Reviewer: ```suggestion
You can convert `WGS84` coordinates to two grids, the `Mapcolonies Control Grid` and `US Army MGRS`. In order to choose your target grid you pass the `target_grid` query parameter.<br/>
```

curl --location '<geocoding_url>/lookup/coordinates?lat=52.57326537485767&lon=12.948781146422107&target_grid=control' \
--header 'x-api-key: <x-api-key>' \
--header 'x-user-id: <x-user-id>'
```

```json title="Response"
{
    "type": "Feature",
    "geocoding": {
        "version": "0.1.0",
        "query": {
            "lat": 52.57326537485767,
            "lon": 12.948781146422107,
            "target_grid": "control"
        },
        "response": {
            "max_score": 1,
            "results_count": 1,
            "match_latency_ms": 0
        }
    },
    "bbox": [ 12.93694771534361, 52.51211561266182, 13.080296161196031, 52.60444267653175 ],
    "geometry": {
        "type": "Polygon",
        "coordinates": [ [ [ 12.93694771534361, 52.51211561266182 ],
                [ 12.93694771534361, 52.60444267653175 ],
                [ 13.080296161196031, 52.60444267653175 ],
                [ 13.080296161196031, 52.51211561266182 ],
                [ 12.93694771534361, 52.51211561266182 ] ] ]
    },
    "properties": {
        "name": "BRN",
        "tileName": "BRN",
        "subTileNumber": [ "06", "97", "97" ]
    }
}
```
</details>

## Common Query Params

> Almost all of our routes consists the same common query params: geo_context, geo_context_mode, limit & disable_fuzziness. <br/>

Reviewer: ```suggestion
Almost all of our routes consists of the same common query parameters: `geo_context`, `geo_context_mode`, `limit` and `disable_fuzziness`. <br/>
```

> * `geo_context`: via this param you can give the search engine the geo context of the search.

Reviewer: ```suggestion
* `geo_context`: via this param you can provide the search engine for geo context of the search.
```

> * `geo_context_mode`: via this param you tell the search engine what to do with `geo_context`. You can filter results (which will result with every feature that matches the query and intersects with `geo_context` shape) or you can bias the results. So features that intersects with `geo_context` will appear first.

Reviewer: ```suggestion
* `geo_context_mode`: via this param you tell the search engine what to do with `geo_context`. You can filter results (which will result with every feature that matches the query and intersects with `geo_context` shape) or you can bias the results. So features that intersect with the `geo_context` will appear first.
```

> * `limit`: by default, we will return our top 5 features that matches the query. You can change the limit and set it anything from 1 to 15 maximum results. We might return less than `limit`, but we limit the returned amount.

Reviewer: ```suggestion
* `limit`: by default, we will return our top 5 features that match the query. You can change the limit and set it from 1 to 15 maximum results. If there are few results, the response may contain less than `limit`, but the importance is limiting the maximum returned values.
```

> * `disable_fuzziness`: fuzziness is on by default. If you want exact match, you may set `disable_fuzziness: true`. 

Reviewer: ```suggestion
* `disable_fuzziness`: fuzziness is on by default. If you want an exact match you may set `disable_fuzziness: true`. 
```

## Usage

> :::caution

Reviewer: Please add a link to the explanation on `tokens`.

Author: Done

**You will need an API token as part of the service authentication &nbsp;**
:::

:::caution

> **All routes require `x-api-key` and `x-user-id`.<br/> 

Reviewer: After adding the link it isn't needed to explain this in detail (for the token).
